src/hb-aat.h

```diff
@@ -0,0 +1,765 @@
+  HB_AAT_LAYOUT_SELECTOR_STYLISTIC_ALT_SIXTEEN_ON    = 32,
+  HB_AAT_LAYOUT_SELECTOR_STYLISTIC_ALT_SIXTEEN_OFF   = 33,
+  HB_AAT_LAYOUT_SELECTOR_STYLISTIC_ALT_SEVENTEEN_ON  = 34,
+  HB_AAT_LAYOUT_SELECTOR_STYLISTIC_ALT_SEVENTEEN_OFF = 35,
+  HB_AAT_LAYOUT_SELECTOR_STYLISTIC_ALT_EIGHTEEN_ON   = 36,
+  HB_AAT_LAYOUT_SELECTOR_STYLISTIC_ALT_EIGHTEEN_OFF  = 37,
+  HB_AAT_LAYOUT_SELECTOR_STYLISTIC_ALT_NINETEEN_ON   = 38,
+  HB_AAT_LAYOUT_SELECTOR_STYLISTIC_ALT_NINETEEN_OFF  = 39,
+  HB_AAT_LAYOUT_SELECTOR_STYLISTIC_ALT_TWENTY_ON     = 40,
+  HB_AAT_LAYOUT_SELECTOR_STYLISTIC_ALT_TWENTY_OFF    = 41
+};
+
+
+/*
+ * Selectors for HB_AAT_LAYOUT_FEATURE_TYPE_CONTEXTUAL_ALTERNATIVES
+ *
+ * Since: REPLACEME
+ */
+enum
+{
+  HB_AAT_LAYOUT_SELECTOR_CONTEXTUAL_ALTERNATES_ON        = 0,
+  HB_AAT_LAYOUT_SELECTOR_CONTEXTUAL_ALTERNATES_OFF       = 1,
+  HB_AAT_LAYOUT_SELECTOR_SWASH_ALTERNATES_ON             = 2,
+  HB_AAT_LAYOUT_SELECTOR_SWASH_ALTERNATES_OFF            = 3,
+  HB_AAT_LAYOUT_SELECTOR_CONTEXTUAL_SWASH_ALTERNATES_ON  = 4,
+  HB_AAT_LAYOUT_SELECTOR_CONTEXTUAL_SWASH_ALTERNATES_OFF = 5
+};
+
+
+/*
+ * Selectors for HB_AAT_LAYOUT_FEATURE_TYPE_LOWER_CASE
+ *
+ * Since: REPLACEME
+ */
+enum
+{
+  HB_AAT_LAYOUT_SELECTOR_DEFAULT_LOWER_CASE      = 0,
+  HB_AAT_LAYOUT_SELECTOR_LOWER_CASE_SMALL_CAPS   = 1,
+  HB_AAT_LAYOUT_SELECTOR_LOWER_CASE_PETITE_CAPS  = 2
+};
+
+
+/*
+ * Selectors for HB_AAT_LAYOUT_FEATURE_TYPE_UPPER_CASE
+ *
+ * Since: REPLACEME
+ */
+enum
+{
+  HB_AAT_LAYOUT_SELECTOR_DEFAULT_UPPER_CASE      = 0,
+  HB_AAT_LAYOUT_SELECTOR_UPPER_CASE_SMALL_CAPS   = 1,
+  HB_AAT_LAYOUT_SELECTOR_UPPER_CASE_PETITE_CAPS  = 2
+};
+
+
+/*
+ * Selectors for HB_AAT_LAYOUT_FEATURE_TYPE_CJK_ROMAN_SPACING_TYPE
+ *
+ * Since: REPLACEME
+ */
+enum
+{
+  HB_AAT_LAYOUT_SELECTOR_HALF_WIDTH_CJK_ROMAN    = 0,
+  HB_AAT_LAYOUT_SELECTOR_PROPORTIONAL_CJK_ROMAN  = 1,
+  HB_AAT_LAYOUT_SELECTOR_DEFAULT_CJK_ROMAN       = 2,
+  HB_AAT_LAYOUT_SELECTOR_FULL_WIDTH_CJK_ROMAN    = 3
+};
+
+
+/**
+ * hb_aat_layout_feature_record_t:
+ *
+ * Feature record and
+ *
+ * Since: REPLACEME
+ **/
+typedef struct hb_aat_layout_feature_record_t
+{
+  hb_aat_layout_feature_type_t feature;
+  hb_ot_name_id_t name_id;
+} hb_aat_layout_feature_record_t;
+
+HB_EXTERN unsigned int
+hb_aat_layout_get_features (hb_face_t                      *face,
+			    unsigned int                    start_offset,
+			    unsigned int                   *record_count, /* IN/OUT.  May be NULL. */
+			    hb_aat_layout_feature_record_t *record_buffer /* OUT.     May be NULL. */);
+
+/**
+ * hb_aat_feature_t:
+ *
+ * Feature value
+ *
+ * Since: REPLACEME
+ */
+typedef unsigned int hb_aat_layout_feature_setting_t;
+
+/**
+ * hb_aat_layout_feature_type_selector_t:
+ *
+ * Feature type selector
+ *
+ * Since: REPLACEME
+ **/
+typedef struct hb_aat_layout_feature_type_selector_t
+{
+  hb_aat_layout_feature_setting_t setting;
+  hb_ot_name_id_t name_id;
+} hb_aat_layout_feature_type_selector_t;
```

Comment:
"feature type selector" as the name here, where does it come from?  What does it mean?  I would have gone with "feature entry" or something.

Reply:
Or better yet.  Remove it.  Add API that just returns feature_setting_t, and separate API to get name_id of a feature, if that makes sense / not too slow / can bsearch?

Reply:
Now both are using bsearch but to be honest I can't see what is guaranteed and what now.

`The setting name arrays must follow the feature name arrays. Each setting name array is sorted by setting. However, the setting name arrays can be arranged in any order.

The feature name array follows the feature name header. The feature name array is sorted by feature type. A featureName record specifies the feature ID, the name table index for the feature's name, and the offset in bytes to the feature's name in the name table. The format of the feature name array is as follows:`